proposals/1125_dynamic-supervisor-oidc-clients/README.md

```diff
@@ -0,0 +1,531 @@
+  - jsonPath: '{range .spec.allowedScopes[?(@ == "pinniped:request-audience")]}{true}{end}{false}'
+    name: Privileged
+    type: boolean
+  - jsonPath: .status.phase
+    name: Status
+    type: string
+  - jsonPath: .status.totalClientSecrets
+    name: Total
+    type: integer
+  - jsonPath: .metadata.creationTimestamp
+    name: Age
+    type: date
+```
+
+##### Configuring client secrets
+
+We wish to avoid storage of client secrets (passwords) in plaintext. They must be stored encrypted or hashed and must
+be generated by the server.
+
+Perhaps the most common approach for this is to use [bcrypt](https://en.wikipedia.org/wiki/Bcrypt) with a random salt
+and a sufficiently high input cost. The salt protects against rainbow tables, and the input cost provides some
+protection against brute force guessing when the hashed password is leaked or stolen. However, the input cost also makes
+it slower for users to authenticate. The cost must be balanced against the current compute power available to attackers
+versus the inconvenience to users caused by a long pause during a genuine login attempt.
+
+Many OIDC Providers auto-generate client secrets and return the generated secret once (and only once) in their API or
+UI. This is good for ensuring that the secret contains a large amount of entropy by auto-generating long random strings
+using lots of possible characters.  We will follow this approach.
+
+Even if the client secrets are hashed with bcrypt, the hashed value is still very confidential, due to the opportunities
+for brute forcing provided by knowledge of the hashed value. Confidential data in Kubernetes should be stored in Secret
+resources. This makes it explicit that the data is confidential.
+[Kubernetes best practices suggest](https://kubernetes.io/docs/concepts/configuration/secret/#information-security-for-secrets)
+that admins should use authorization policies to restrict read permission to Secrets as much as possible. Additionally,
+some clusters may use the Kubernetes feature to
+[encrypt Secrets at rest](https://kubernetes.io/docs/tasks/administer-cluster/encrypt-data/), and thus reasonably expect
+that all confidential data is encrypted at rest.  We will use Kubernetes secrets to store the client secret hash.
+
+CRDs are convenient to use, however, they have one core limitation - they cannot represent non-CRUD semantics.  For
+example, the existing token credential request API could not be implemented using a CRD - it processes an incoming token
+and returns a client certificate without writing any data to etcd.  Aggregated APIs have no such limitation.  We will
+use an aggregated API to handle generation of client secrets.
+
+We will use the existing token credential request API as a model for the new OIDC client secret request API.  Only the
+`create` verb will be supported (but this resource will be part of the `pinniped` category and will have no-op `list`
```

Comment:
While I think this may be fine for an mvp, I also think being able to use `kubectl delete` would make the experience a lot nicer and more intuitive for users who are used to kubectl based apis and I think we should prioritize that work soon.

Reply:
Agree, and nothing prevents us from building that stuff in the future.  There is also some value and waiting for feedback.